crates/shared/src/caching_balance_fetcher.rs

```diff
@@ -0,0 +1,265 @@
+use crate::{
+    account_balances::{BalanceFetching, Query, TransferSimulationError},
+    current_block::{into_stream, CurrentBlockStream},
+};
+use anyhow::Result;
+use futures::StreamExt;
+use itertools::Itertools;
+use model::order::SellTokenSource;
+use primitive_types::{H160, U256};
+use std::{
+    collections::HashMap,
+    sync::{Arc, RwLock},
+};
+use tracing::Instrument;
+
+#[derive(Default)]
+struct BalanceCache {
+    /// Block number the cached data was collected from.
+    block: u64,
+    data: HashMap<Query, U256>,
+}
+
+pub struct CachingBalanceFetcher {
+    inner: Arc<dyn BalanceFetching>,
+    balance_cache: Arc<RwLock<BalanceCache>>,
+}
+
+impl CachingBalanceFetcher {
+    pub fn new(inner: Arc<dyn BalanceFetching>) -> Self {
+        Self {
+            inner,
+            balance_cache: Default::default(),
+        }
+    }
+}
+
+struct CacheResponse {
+    // The indices and results of queries that were in the cache.
+    cached: Vec<(usize, Result<U256>)>,
```

Comment:
Seems like this could be without Result because it is never set to Err.

Reply:
This was done to avoid creating yet another vector to merge the cached and requested values together.
But I'm fine with changing that.

Reply:
If there is a good reason for it then I don't mind. Would maybe double check that that is really needed and can't be avoided with iterators or something like that.